From 10f4a28914f45a94c4b0efc48bd10105bafa11cd Mon Sep 17 00:00:00 2001
From: Piyush Shah <piyush.shah@espressif.com>
Date: Wed, 22 Nov 2023 18:52:57 +0530
Subject: [PATCH] app_wifi: Add an option to provide a custom Proof of
 Possession (PoP) for provisioning

POP_TYPE_CUSTOM has been added as an option for app_wifi_start().
app_wifi_set_custom_pop() should be used before this.
---
 examples/common/app_wifi/app_wifi.c | 27 +++++++++++++++++++++++++++
 examples/common/app_wifi/app_wifi.h |  7 ++++++-
 2 files changed, 33 insertions(+), 1 deletion(-)

diff --git a/examples/common/app_wifi/app_wifi.c b/examples/common/app_wifi/app_wifi.c
index c9cd770..c7ed07d 100644
--- a/examples/common/app_wifi/app_wifi.c
+++ b/examples/common/app_wifi/app_wifi.c
@@ -255,6 +255,27 @@ static esp_err_t read_random_bytes_from_nvs(uint8_t **random_bytes, size_t *len)
     return ESP_ERR_NO_MEM;
 }
 
+static char *custom_pop;
+esp_err_t app_wifi_set_custom_pop(const char *pop)
+{
+    /* NULL PoP is not allowed here. Use POP_TYPE_NONE instead. */
+    if (!pop) {
+        return ESP_ERR_INVALID_ARG;
+    }
+
+    /* Freeing up the PoP in case it is already allocated */
+    if (custom_pop) {
+        free(custom_pop);
+        custom_pop = NULL;
+    }
+
+    custom_pop = strdup(pop);
+    if (!custom_pop) {
+        return ESP_ERR_NO_MEM;
+    }
+    return ESP_OK;
+}
+
 static esp_err_t get_device_service_name(char *service_name, size_t max)
 {
     uint8_t *nvs_random = NULL;
@@ -279,6 +300,12 @@ static char *get_device_pop(app_wifi_pop_type_t pop_type)
 {
     if (pop_type == POP_TYPE_NONE) {
         return NULL;
+    } else if (pop_type == POP_TYPE_CUSTOM) {
+        if (!custom_pop) {
+            ESP_LOGE(TAG, "Custom PoP not set. Please use app_wifi_set_custom_pop().");
+            return NULL;
+        }
+        return custom_pop;
     }
     char *pop = calloc(1, POP_STR_SIZE);
     if (!pop) {
diff --git a/examples/common/app_wifi/app_wifi.h b/examples/common/app_wifi/app_wifi.h
index 9c2d331..58c0704 100644
--- a/examples/common/app_wifi/app_wifi.h
+++ b/examples/common/app_wifi/app_wifi.h
@@ -48,12 +48,17 @@ typedef enum {
     /** Do not use any PoP.
      * Use this option with caution. Consider using `CONFIG_APP_WIFI_PROV_TIMEOUT_PERIOD` with this.
      */
-    POP_TYPE_NONE
+    POP_TYPE_NONE,
+    /** Use a custom PoP.
+     * Set a custom PoP using app_wifi_set_custom_pop() first.
+     */
+    POP_TYPE_CUSTOM
 } app_wifi_pop_type_t;
 
 void app_wifi_init();
 esp_err_t app_wifi_start(app_wifi_pop_type_t pop_type);
 esp_err_t app_wifi_set_custom_mfg_data(uint16_t device_type, uint8_t device_subtype);
+esp_err_t app_wifi_set_custom_pop(const char *pop);
 
 #ifdef __cplusplus
 }
-- 
2.37.3

